### ISSUE TITLE
Gintro failed to install on Debian Buster without Gtk4

### ISSUE THREAD

**Issue by 1jss** [StefanSalewski/gintro#176]
New to the nim ecosystem I tried to install Gintro on nim 1.4.8 by running `nimble install gintro` (nimble v0.13.1) but got the following errors:

    nimscript.nim(260, 7)
    Error: unhandled exception: FAILED: /tmp/gintrosalewski/gen 1 [OSError]
    Error: Exception raised during nimble script execution

None of the gtk4 packages are installed, but `libgtk-3-dev` is. (I have GI working in Python 3)

I also noticed the following in the verbose output:

    /tmp/gintrosalewski/gen.nim(3817) main
    SIGSEGV: Illegal storage access. (Attempt to read from nil?)

**Comment by StefanSalewski** [comment]
Thanks for reporting. Seems to be again a very strange error, which was reported two more times already, one by a Debian stable, and by an very very old Ubuntu. See

https://github.com/StefanSalewski/gintro/issues/166


 Our feeling is that it is a gobject-introspection issue, so not related to Nim and gintro. I don't know what Buster is -- some decades ago we had in Germany a porn actress called Dolly Buster, but I guess it is not really related :-) 

I have the feeling that you are very new to Nim and GTK, and maybe to computer programming in general. So I would strongly recommend not to using GTK/gintro, but one of the other 20 Nim GUI toolkits. I have listed them at the start of the Nim GTK4 book. Mr Rumpf and some other Nim core devs seems to favour the fidget GUI toolkit currently, so maybe you should start with this one. GTK is really hard to learn, it is not very popular for Windows and Mac users, and has no Android and Web support.

Maybe in some years, when you have some more Nim experience, you can come back to this issue and help us to find the source for the error. For me it would be necessary to install such an old Debian Linux to investigate the issue, which is a lot of work, and I may only proof that it is an gobject-introspection issue, which I then still may not fix myself.

**Comment by 1jss** [comment]
Thanks for your reply and for the link! I'll go on from there.

Buster is the version name for Debian ([Link](https://www.debian.org/releases/buster/)). It's the old stable release, so Gtk4 is not in the repos. The name "Buster" comes from a character in the Toy Story movie.

Partly right guessed. I'm completely new to Nim, but not to programming in general or Gtk. I'm coming from Gtk3 development in Python3 and now want to see if Nim can be a feasible language to learn and port my Python + Gtk3 apps to. I found your Gtk Programming book and think I can find my way around it if I can get gintro to work. I installed `oldgtk3` without problems, but would like to use the GI bindings instead, like I've done in Python.

**Comment by StefanSalewski** [comment]
>but not to programming in general or Gtk.

OK, then gintro is indeed an option for you. 

>It's the old stable release

That would confirm my guess that it is an gobject-introspection issue.

Is there a chance that you can install a fresh OS? Very old GTK, that is older than about 18 months, would have one more issue that you would discover after successful install: It is related to cairo, which does not really support gobject-introspection. About one year ago we decided to split the cairo modules into one base module and one hand crafted module. That split is not supported by very old GTK (older than 18 months). I would think that no Linux user should still use such an old GTK still, at least not in the free world. Maybe in north korea, china or russia?

One solution would be that we send you the generated gtk modules. We have not tested that before. At least you would have to tell us if you have a 32 or a 64 bit OS, and if it is ARM, x86 or what else. Then still our shipped files would contain all the latest GTK functions, which may be not available on your box. And maybe library versions would not really match, but when you have some experience with computers, you could patch that with an editor.

Let us know when you need more help.

**Comment by 1jss** [comment]
Yes, I can set it up in a vm. The version of Gtk is 3.24.5 and have had no problems with Gtk3 (through gi) in Python3 on this machine.

Interesting with the Cairo split. What would be the last version of Gintro that supported older versions of Gtk3?

I'm on x86_64 and would be willing to try the generated modules.

My problem seems to be an exact duplicate of https://github.com/StefanSalewski/gintro/issues/151

**Comment by StefanSalewski** [comment]
For Python gobject-introspection use always only the tiny part that is necessary to execute the actual Python script, while gintro generates Nim modules which contain every data type and every function available. This way we have found some bugs in GIR files, which had existed for decades undiscovered. The GTK team has fixed that bugs now.

For the cairo split, I can not remember when it was done, but generally I would not recommend using an older gintro version, as older versions had some bugs. The cairo splitb issue was easy to fix manually by just ignoring cairo, one user did that. And when you should use the pregenerated files cairo would just work for you.

As I am using 64 bit x86 too, I can indeed try to make the pregenerated files available for you. One question is how you can install the files so that the compiler can  find them. I have always done the install with nimble, but I do not know what nimble is doing in detail.

Can you download files from a German webpage from your country? I am not sure for China and Russia -- Iran and North Korea seems to block file exchange unfortunately.

Tomorrow I can put a tar archive on my homepage containing all the files, will post the link here then.

```$ du -h ~/.nimble/pkgs/gintro-#v0.9.2/gintro
15M	/home/salewski/.nimble/pkgs/gintro-#v0.9.2/gintro

```

So it would be 15 MBytes in size, is that OK for you. Or can I compress the files? can you uncompress gzip? I hope you have not still to use a 2400 baud modem for download.

**Comment by 1jss** [comment]
Awesome! Yes, the files won't even have to leave the EU, so you don't have to compress them further.

**Comment by 1jss** [comment]
So here's some more data:

I managed to install v0.8.2, but the subsequent versions fail. Trying to compile a simple program with this version gives me:

`cairoimpl.nim(653,22) Error: undeclared identifier: 'Rectangle'`

Which is exactly what you mentioned above, the cairo library bindings.

**Comment by StefanSalewski** [comment]
That is interesting, I had assumed that older gintro versions would fail too during install when gobject-introspection is broken. But maybe gobject-introspection is broken only for some modules, which are not available in older gintro. To fix above error it should be sufficient to edit file

```$ grep -i "include cairoimpl" ~/.nimble/pkgs/gintro-#v0.9.2/gintro/*
/home/salewski/.nimble/pkgs/gintro-#v0.9.2/gintro/cairo.nim:include cairoimpl
```

that is for your current  v0.8.2

gedit ~/.nimble/pkgs/gintro-#v0.8.2/gintro/cairo.nim


and remove or comment out the line with the 

include cairoimpl

statement.

Then most of the GTK3 examples should compile and work. Exceptions would be the examples that indeed make use of the cairo lib, that is the drawingarea.nim example and some related examples.

To investigate in more detail why install  of v0.9.4 fails on your box, it would make sense to look at the install log and see which module was created last successfully. Maybe it fails for the webkitgtk, gstreamer, libhandy, vte, or similar less important modules. In the past some people asked for that modules, so I added them. But actual number of users in of course tiny, you would not need them. So in principle you could fix the gen.nim file which at the bottom has listed all the modules to use. But as a beginner you will not know how you can interrupt the nimble install process to modify the gen.nim file and then continue. I don't know myself, for development I am using a handmade directory which I created many years ago, in which I have collected all the files I need to compiles and test run the gen.nim script. If you can access NIm IRC or the Nim Forum people may tell you. 

Are you familiar with git and can you use github from your country? In that case you could fork gintro v0.9.4 and comment out all the unneeded modules, and then you would do the nimble install from your repository giving nimble the full URI.

If you are not able to do this, then I may temporary patch latest head version to not generate all the exotic modules, and patch it back after you have successfully installed. You would to have then only to remove the above mentioned include directive and it should work for you.

I think these suggestions would be better when they should work as shipping pregenerated files, as with pregenerated files we have absolutely no experience, so we do not know what can go wrong.

Let me know your success.

**Comment by StefanSalewski** [comment]
For the actual file patch of gen.nim

```
  else:
    echo "First we try generating bindings for GTK4, this may fail when GTK4 is not properly installed"
    echo "on your computer. But don't worry, you can still use GTK3"
    main("Gtk", "4.0") # the 3 new 4.0 releases
```

you could add a discard and comment out all the indented statements after the else like

```
  else:
    discard
    # echo "First we try generating bindings for GTK4, this may fail when GTK4 is not properly installed"
    # echo "on your computer. But don't worry, you can still use GTK3"
    # main("Gtk", "4.0") # the 3 new 4.0 releases
```

then it would try to install GTK3 only. But GTK3 may fail still, so you would see in the logs which module wasd sucessfully last, and comment out the next one, or maybe all for which you have no use like

```
  if ISGTK3:
    echo "Generating bindings for GTK3..."
    main("Gtk", "3.0") # the 3 new 4.0 releases in old version
    main("Gdk", "3.0")
    main("GdkX11", "3.0")
    # main("Gsk") # not available for GTK3
    ### main("Graphene")
    main("GObject") # and the old common onces. GObject before Glib!
    main("GLib")
    main("Gio")
    main("GdkPixbuf")
    main("GModule")
    ### main("GtkSource", "4")
    main("Atk")
    main("Pango")
    main("PangoCairo")
    main("PangoFT2")
    main("fontconfig")
    main("freetype2")
    main("HarfBuzz")
    main("Rsvg")
    main("xlib")
    ### main("Vte")
    ### main("Notify")
    #### main("Handy")
    ### main("Nice")
    main("cairo")
    ### main("WebKit2", "4.0")
    #### main("JavaScriptCore", "4.0")
    #### main("Soup")
    ### main("WebKit2WebExtension", "4.0")
    ### main("Gst")
    ### main("GstBase")
    ### main("GstAllocators")
    ### main("GstApp")
    ### main("GstAudio")
    ### main("GstCheck")
    ### main("GstController")
    ### main("GstGL")
    ### main("GstInsertBin")
    ### main("GstMpegts")
    ### main("GstNet")
    ### main("GstPlayer")
    ### main("GstRtp")
    ### main("GstRtsp")
    ### main("GstSdp")
    ### main("GstTag")
    ### main("GstVideo")
    ### main("GstWebRTC")
    ### main("GstApp")
    ### main("GstPbutils")
    ### main("GtkLayerShell")
  else:
```

**Comment by 1jss** [comment]
Yes, that's a way better idea! I'll fork the repo, modify and install from there. Thanks for all your help. I'll keep you updated.

**Comment by StefanSalewski** [comment]
When you manage the forking, there is even a simpler solution to fully suppress the gtk4 module generation:

Comment out in  gintro/gintro.nimble the line

exec(td / wd / "gen 1")

**Comment by StefanSalewski** [comment]
Instructions without forking: Looking at

https://github.com/nim-lang/nimble

I get the feeling that we can replace

nimble install gintro

by

nimble develop gintro

modily the files and finally do

nimble build gintro

But I never did that myself before.

**Comment by StefanSalewski** [comment]
And I have just uploaded my actual v0.9.4 files:

http://ssalewski.de/tmp/gintro_0.9.4_files.tar

You should be able to download the tar archive with firefox browser or another tool and then you can extract the files with

```
tar xf gintro_0.9.4_files.tar
```

Maybe at least cairo.nim and cairoimpl.nim is useful for you to get cairo working.

Most of the files have a line with the library version like

```
$ head cairo.nim 
# dependencies:
# immediate dependencies:
# libraries:
# libcairo-gobject.so.2
{.warning[UnusedImport]: off.}

const Lib = "libcairo-gobject.so.2"
{.pragma: libprag, cdecl, dynlib: Lib.}

from gobject import boxedFree
```

which you may have to adjust. Beside that I think that the GTK3 files may work for you, as GTK3 has not really changed in the last years. For GTK4 it is very different, GTK4 function signatures have changes a lot still in the last months. When you get your own install process working somehow, it may be enough to just copy cairo.nim and cairoimpl.nim into the directory with your other gintro modules. And when you get only 0.8.2 to install, then I would recommend to replace all the v0.8.2 files with the files from the tar archive. My cairo lib is cairo-1.16.0 so there may be some tiny differences, but I think cairo API has not changed much in the last years.

**Comment by 1jss** [comment]
Great! I had no problem cloning the repo, and diong the modifications locally i `cd` to the folder and run `nimble install` to install that version.

The install failed on GObject, but commenting that out takes me through the install without errors.

**Comment by StefanSalewski** [comment]
>The install failed on GObject,

That is interesting, but sad for you, as Gobject is the most basic module.

So you would have to take gobject.nim, cairo.nim and cairoimpl.nim from my uploaded files.

I hope it works then all for you!

**Comment by StefanSalewski** [comment]
But your hints that the install works for you with v0.8.2 and that gobject fails for v0.9.4 would enable us to find the reason for the issue. The core issue seems to a gobject-introspection bug, but current v0.9.4 somehow triggers it, but old v.8.2 not. I will look at the code of gen.nim later, maybe I will get an idea. But unfortunately I can not test it. For you copying the 3 files should work for now -- maybe later when you have more experience with Nim you can debug this issue.

**Comment by 1jss** [comment]
It does indeed work!

Thank you so much for all your work!

**Comment by gavr123456789** [comment]
It's good to see someone who already has GTK experience here. I advise you to switch to a newer distribution or even a Rolling Release, and use GTK 4. After all, then the whole GTK 3 that you write will turn into a legacy, there is no point in starting a GTK 3 application now.  
More GTK 4 examples there https://github.com/StefanSalewski/gintro/pull/122

**Comment by 1jss** [comment]
@gavr123456789 While you have a valid point, I actually think there is an underrated value in targeting older OS versions and libraries. Not only is it easier to target something that has a stable API, but also a requirement if you want to target users that are outside of the arch/gentoo bubble, as many distros are not rolling releases. (I know this has partly changed with flatpaks, though.)

I'll take the Elementary OS project as an example. Up until August this year their distro was based on Ubuntu 18.04 LTS, which is supported by Canonical until April 2028. That is a more than 3 year old OS which will have users for many years to come.

**Comment by 1jss** [comment]
Back to the issue.

The build (`nimble install gintro`) seems to halt at `GObject` despite `gobject-introspection` being installed. Might be that it's version (1.58.3-2) is too old. Could be an improvement to the docs to state minimum required version of the libraries.

These are the modules that I tried, and compiled without warnings on Debian 10:
- gtk
- gdk
- gdkx11
- glib
- gio
- gdkpixbuf
- gmodule
- atk
- pango
- pangocairo
- pangoft2
- fontconfig
- freetype2
- harfbuzz
- rsvg
- xlib

Cairo didn't interrupt the build, but using it to compile an app threw the following error:
```.nimble/pkgs/gintro-0.9.4/gintro/cairoimpl.nim(700, 22) Error: undeclared identifier: 'Rectangle'
```
Could this be because of the missing GObject?

I used the preconfigured `cairo.nim`, `cairoimpl.nim` and `gobject.nim` from @StefanSalewski  (by adding them to my `.nimble/pkgs/gintro-0.9.4/gintro/´ folder) and now the gtk3 example applications (that I've tried) have compiled and run without problems.

I didn't have to update the Cairo version number as the same version was installed on my system.

**Comment by 1jss** [comment]
> The core issue seems to a gobject-introspection bug, but current v0.9.4 somehow triggers it, but old v.8.2 not.

On a quick comparison of v0.9.4 and v0.8.2 `gen.nim` I noticed that `main("GLib")` was run before `main("GObject")` in v0.8.2 and not after as in v0.9.4, which was also stated in a comment: `GObject before Glib!`

**Comment by StefanSalewski** [comment]
Thanks, that was my guess too. But I have no real idea why the sequence does matter at all, and for newer gobject-introspection it seems to be important that GObject is createdc before Glib as from the comment. Maybe that has change somehow, I have to think about that, maybe I can ask Mr Bassi. Good to know, we can create a patch for old gobject-introspection!

For the cairo rectangle issue: As far as I can remember, initially in 2015 cairo was not supported at all by gobject-introspection.  So we created a cairo.nim file manually, which is much work and is error prone. Later cairo got some minimal gobject-introspection support for basic types. So we split cairo in a cairo.nim which contains basic types, and cairoimpl.nim with all the rest manually created. So it is not really related to the gobject module issue. But we have indeed two issues for very old gobject-introspection, the cairo one and the gobject one. We can patch both somehow, maybe we will do that.

For the GTK4 vs Gtk3 decision: You are right, it makes some sense that GTK3 is still used for existing apps. But stable GTK4 is now available for one years, so for new apps I would really try to use GTK4 when possible. Some GTK core devs may point you just to GTK4 when you try to ask questions for legacy GTK3 on Forum or IRC, and I would guess that users would not like to install a new app that uses GTK3 still. And even for users that have a distro that has no GTK4, there are solutions. I have installed GTK4 in the last two years from git sources to /opt, which is really not complicated. The exception is webkitgtk2, but we do not need it.

Here is my bash script to install GTK4:

```
$ cat fullgtk4install 
# https://discourse.gnome.org/t/installing-gtk4-for-testing-on-opt-ii/3349/4
cd
rm -rf /opt/gtk
rm -rf gtk
rm -rf gtksourceview
rm -rf vte
rm -rf libhandy
rm -rf libnice
rm -rf pango
rm -rf dconf
rm -rf gstreamer
rm -rf gtk-layer-shell

cd
git clone https://gitlab.gnome.org/GNOME/pango.git
cd pango
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.gnome.org/GNOME/dconf.git
cd dconf
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.gnome.org/GNOME/gtk.git
cd gtk
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.gnome.org/GNOME/vte.git
cd vte
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.gnome.org/GNOME/libhandy.git
cd libhandy
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.freedesktop.org/libnice/libnice.git
cd libnice
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

#cd
#git clone https://gitlab.freedesktop.org/gstreamer/gstreamer.git
#cd gstreamer
#meson --prefix /opt/gtk builddir
#ninja -C builddir
#ninja -C builddir install

export GI_TYPELIB_PATH=/opt/gtk/lib64/girepository-1.0
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/gtk/lib64/
#export GSETTINGS_SCHEMA_DIR=/opt/gtk/share/glib-2.0/schemas /opt/gtk/bin/gtk4-demo
export PKG_CONFIG_PATH="/opt/gtk/lib64/pkgconfig/"

cd
git clone https://gitlab.gnome.org/GNOME/gtksourceview.git
cd gtksourceview
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://gitlab.gnome.org/GNOME/libadwaita.git
cd libadwaita
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

cd
git clone https://github.com/wmww/gtk-layer-shell.git
cd gtk-layer-shell
meson --prefix /opt/gtk builddir
ninja -C builddir
ninja -C builddir install

#cd
#unxz webkitgtk-2.32.1.tar.xz
#tar -xf webkitgtk-2.32.1.tar
#export PKG_CONFIG_PATH="/opt/gtk/lib64/pkgconfig/"
#cd webkitgtk-2.32.1
#cmake -DUSE_GTK4=ON -DENABLE_BUBBLEWRAP_SANDBOX=OFF -DCMAKE_INSTALL_PREFIX=/opt/gtk -DPORT=GTK -DCMAKE_BUILD_TYPE=RelWithDebInfo -#GNinja
#ninja
#ninja install
```

To use that together with regular GTK3:

```
export GI_TYPELIB_PATH=/opt/gtk/lib64/girepository-1.0
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/gtk/lib64/
export GSETTINGS_SCHEMA_DIR=/opt/gtk/share/glib-2.0/schemas /opt/gtk/bin/gtk4-demo
export PKG_CONFIG_PATH="/opt/gtk/lib64/pkgconfig/"
```

Then I can do "nimble install gintro" and get GTK3 and GTK4 support. For running GTK4 apps

export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/gtk/lib64/

is all we need later.

But now my Gentoo distribution finally got some GTK4 support also -- libadwaita is still missing and webkitgtk does not support GTK4 by Gentoo.

Generally I think that people interested in Nim are interested in up to date software, so they would be interested in GTK4 more than in GTK3. At least when they install from sources. When you intent to distribute binary GTK3 apps them it may be different.

Besides installing GTK4 from git sources, there are more option of course, flatpak and all that, and Gnome has even shipped something like a GnomeOS to support GTK4. But I know no details.

**Comment by 1jss** [comment]
I'm trying to get some printouts of what happens just before it crashes on line 3817 where we run ` gStructInfoFindMethod(info, n)`

Info is `gBaseInfoGetName(info)`
Action is the iterator `n`

    ...
    Info:EnumValue
    Action:unref
    Info:EnumValue
    Action:
    Info:EnumValue
    Action:get_qdata
    Info:FlagsClass
    Action:free
    Info:FlagsClass
    Action:unref
    Info:FlagsClass
    Action:
    Info:FlagsClass
    Action:get_qdata
    Info:FlagsValue
    Action:free
    Info:FlagsValue
    Action:unref
    Info:FlagsValue
    Action:
    Info:FlagsValue
    Action:get_qdata
    Info:InterfaceInfo
    Action:free
    Info:InterfaceInfo
    Action:unref
    Info:InterfaceInfo
    Action:
    Info:InterfaceInfo
    Action:get_qdata
    Info:ObjectClass
    Action:free

and then the Error:

    Traceback (most recent call last)
    /tmp/gintrosalewski/gen.nim(4312) gen
    /tmp/gintrosalewski/gen.nim(4185) launch
    /tmp/gintrosalewski/gen.nim(3819) main
    SIGSEGV: Illegal storage access. (Attempt to read from nil?)

So the problem seems to arise when we run `gStructInfoFindMethod(ObjectClass, unref)`.

I haven't dug into what this means yet, though.

**Comment by StefanSalewski** [comment]
My feeling is that for ordinary people debugging such types of issues is very hard. The initial creators of gobject-introspection are gone, and there are still a lot of issues. I think when I started with gintro in late 2015 gobject-introspection had 116 open issues. Mr Bassi may have some insight, but he is generally very busy. When asking him we should try to provide C code that shows the issue at least.

Glib and gobject are strongly coupled, so order of processing seems to matter. Current order was working with up to date GTK3 and with GTK4, but seems to fail with old libs. I guess changing the order will not work with current libs.

I just had one idea: With some gst modules we noticed that it is necessary to open the libs with dlopen before use, otherwise we got a strange crash. We discussed that with Mr Droege and Mr Bassi on Gnome forum, conclusion was that for libs which provides a init() function we should open them with dlopen() before using gobject-introspection. But it was necessary only really for one single gst module. We may try applying dlopen() to glib and gobject as well. You can find the code for dlopen() in gen.nim, should be not too hard to add glib and gobject. But that is only a guess. Another way may be to process glib and gobject somehow in parallel, not one after the other. For that I would have to inspect the code in gen.nim in detail to see if it may be possible.

**Comment by 1jss** [comment]
Despite the order of inits being a difference between the versions, I doubt that's the actual problem as I changed the order in my local v0.9.4 without getting any further.

I'll try to get some printouts from v0.8.2 as well to see what the data looks like when it works. I migt also try out your idea with the inits. Back in a few days.

Again, thanks for your awesome work with this library and your books!

**Comment by 1jss** [comment]
Regarding `cairo.nim` and `cairoimpl.nim` our files are identical except one type declaration in `cairo.nim` that is completely missing in my build:

    type
      Rectangle* {.pure, byRef.} = object
        x*: cdouble
        y*: cdouble
        width*: cdouble
        height*: cdouble
    
    proc cairo_gobject_rectangle_get_type*(): GType {.importc, libprag.}

As I understand the code this should be added by `cairo.Rectangle` in `manCallerAlloc` on line 494. The odd thing is that `cairo.RectangleInt` on the next line is included without problems. Any ideas?

**Comment by StefanSalewski** [comment]
>compiles without complaints.

That is already an interesting observation. Well maybe.

But generally the logic in gen.nim has to be like it is, otherwise we may get different binding module files.

For above logic expression, you can apply De Morgan's laws to modify it, see https://en.wikipedia.org/wiki/De_Morgan%27s_laws

We have

not ((gStructInfoIsGtypeStruct(info) and n != "ObjectClass"))

The outer parent can be removed of course, and we may write the expression

not (gStructInfoIsGtypeStruct(info) and (n != "ObjectClass"))


Applying De Morgan we get

(not gStructInfoIsGtypeStruct(info) or (n == "ObjectClass"))

As you see, De Morgan makes the and an or!

But what we may use is that "a and b" is the same as "b and a" logically, but that when the first subterm is false the second is not evaluated at all in most programming languages including C and Nim. So we can write the initial expression as

 not ((n != "ObjectClass") and gStructInfoIsGtypeStruct(info))

So gStructInfoIsGtypeStruct(info) is not evaluated at all when the first subexpression is false. Maybe, with much luck, the call gStructInfoIsGtypeStruct(info) generates the crash, and it has not to be called when (n == "ObjectClass").

I hope I get the brackets right, it is a bit difficult to type here....

**Comment by StefanSalewski** [comment]
>our files are identical except one type declaration

That is interesting, I would have expected more missing data types. If you would use an even older GTK version more types would be missing. But when that rectangle is the only one, we may try to apply that data type manually and hope that no one is using an even older GTK.

What is available by gobject-introspection is determined by the so called GIR XML files or their binary typelib versions. You can inspect the GIR files with an text editor:

```
$ locate cairo-1.0.gir
/usr/share/gir-1.0/cairo-1.0.gir
salewski@nuc ~/SDTold $ 
salewski@nuc ~/SDTold $ grep -A24 "record name=\"Rectangle" /usr/share/gir-1.0/cairo-1.0.gir 
    <record name="Rectangle" c:type="cairo_rectangle_t"
            glib:type-name="CairoRectangle"
            glib:get-type="cairo_gobject_rectangle_get_type">
      <field name="x" writable="1">
        <type name="gdouble" c:type="gdouble"/>
      </field>
      <field name="y" writable="1">
        <type name="gdouble" c:type="gdouble"/>
      </field>
      <field name="width" writable="1">
        <type name="gdouble" c:type="gdouble"/>
      </field>
      <field name="height" writable="1">
        <type name="gdouble" c:type="gdouble"/>
      </field>
    </record>
    <record name="RectangleInt" c:type="cairo_rectangle_int_t"
	    glib:type-name="CairoRectangleInt"
	    glib:get-type="cairo_gobject_rectangle_int_get_type">
      <field name="x" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="y" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="width" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="height" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
    </record>
    <function name="image_surface_create" c:identifier="cairo_image_surface_create">
      <return-value transfer-ownership="none">
	<type name="none" c:type="void"/>
      </return-value>
      <parameters>
      </parameters>
    </function>
  </namespace>
</repository>
```

The first record name may be missing in your gir file, or it may be ignored by gobject-introspection. Entries marked with introspectable="0" are ignored.

Note that editing your Gir file would make no difference for you, as gobject-introspection always uses a "compiled " binary variant, called typelib file. You may be able to create typelib files yourself, but you should not try that, as that may damage your GTK install.

**Comment by StefanSalewski** [comment]
The funny fact is, that the location where the crash actually occurred for you is a code section which is not really necessary. That section follows this line:

if namespace notin ["HarfBuzz", "cairo"]: # auto-detecting the "light" entities is too hard, but here we get some candidates

That section was used in the past only to find some candidates for the "light" objects like GtkTextIter or Rectangle and many more. These light symbols are generally allocated on the stack, and we may access fields directly. So these light symbols do not need a Nim proxy object. These symbols generally have no constructor, as they are generally allocated on the Stack. Unfortunately it is not easily possible to identify these light symbols fully automatically by gobject-introspection, I had a discussion with Mr. Droege, the GStreamer and Rust GTK dev about that. For Nim we use a manually generated list of light symbols, but that code can print out some candidates.

For the other two users it was the same, I was aware that the error actually occurred in a code section that is not really necessary. But most users are just so lazy that I just tell them to use a different GUI toolkit. I left that issue open for a long period, so others would have been able to see that issue easily. But then recently I closed a few old stalled issues. And indeed I was only aware of the line number and the fact that that code segment was not needed, but I was not aware for which module that crash occurred, the other two  users did not tell me.

But well, that issue should not occur, I don't think that my code in gen.nim is wrong. And when you remove that section the crash may occur with the same gobject module at another location. When not -- OK then I can remove that section.

A correct investigation would mean to recreate the crashing code in plain C and ask Mr Bassi or Mr Droege  -- there are not many other bright gtk devs left. But both are busy, should we steal their time with an error in such old code that nearly nobody use still. And both devs would have to find an old Debian Buster install to verify the crash.

Finally -- you told me that 0.8.2 does not crash for you, and I could imagine that that code section was introduced after 0.8.2. When we remove it, all may be fine. We only have to add the rectangle fix. So somehow you seem to have solved the issue. We do not know the real reason, but we can avoid it! Thanks.

**Comment by 1jss** [comment]
> (not gStructInfoIsGtypeStruct(info) or (n == "ObjectClass"))

Wow! Boolean math is hard! After I now understand your conversion, but the question is still what we are trying to achieve...

> Maybe, with much luck, the call gStructInfoIsGtypeStruct(info) generates the crash, and it has not to be called when (n == "ObjectClass").

...and this is good information, because that is exactly what happens! I will try to find out why.

**Comment by StefanSalewski** [comment]
>question is still what we are trying to achieve

See my post before, it is just a filter at that position.

We read all the struct candidates, but ignore the class structs I think. So indeed inverting the order of the two subterms would be the better way, it would be faster at least.

**Comment by 1jss** [comment]
> We read all the struct candidates, but ignore the class structs I think

Great! This line should deffinitely be a comment in the source.

**Comment by 1jss** [comment]
I've tried to break down the logic and it does not prevent `gStructInfoIsGtypeStruct(info)` when `n == "ObjectClass"`

    (not gStructInfoIsGtypeStruct(info) or (n == "ObjectClass"))

- a: `IsGtypeStruct`
- b: `ObjectClass`
- c: `not a`
- d: `b or c`

||1|2|3|4|
|-  |-|-|-|-|
|a  |true|true|false|false|
|b  |false|true|false|true|
|c  |false|false|true|true|
|d  |false|true|true|true|

See column 2. The condition returns `true` when `a` and `b` are both `true`.

**Comment by 1jss** [comment]
This is my solution for the build error:

    if not (gStructInfoIsGtypeStruct(info) or (n == "ObjectClass")):

If the conditional statement is meant to filter out `GtypeStruct` and `ObjectClass` this does it, while the current line doesn't. Please try this on Gtk4 and see if the output is changed. If nothing is changed I'll make a pull request.

**Comment by StefanSalewski** [comment]
Thanks.

As it seems that only these collecting of light symbol candidates causes the error, I will apply your rewrite and finally comment out that section fully. It is in no way needed. We introduced it in the hope that we could fully automatically find all the light symbols, but it became clear that that does not work. We leave that code active, but it really does nothing. Maybe for GTK5, when new light symbol candidate arrive, or when we decide to add even more modules with gobject-introspection support we can activate that code again.

For the rectangle fix, I have to decide how we do it best: Generally ignore the Rectangle provided by gobject-introspection and add the type always manually, or scan the gobjection-introspection struct list for cairo, see if it is provided, and only add it manually if missing. Should both work. I will add these fixes in the next days and tell you when you can test again.

**Comment by 1jss** [comment]
Awesome! Thanks for your huge responsiveness!

I'll also look into why my system doesn't find the Rectangle but everything else in the Cairo binding.

**Comment by StefanSalewski** [comment]
The crash may be related to this old bug which Google found.

https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/86

No idea if that one was finally fixed.

**Comment by StefanSalewski** [comment]
Your suggest fix makes indeed sense!

You may try with

```
nimble uninstall gintro
nimble install gintro@#head
```

The fixes are tiny, so I have not created a new version number yet.

```
$ diff gen.nim ~/gintro/tests/gen.nim 
3c3
< # v 0.9.4 2021-SEP-15
---
> # v 0.9.4 2021-AUG-13
44,45d43
< {.warning[CStringConv]: off.}
< 
300,301d297
< var cairoProvidesRectangle = false # ugly fix for legacy Debian Buster, CairoRectangle was missing
< 
452c448
< # These objects are plain structs like GtkTextIter containing only fundamental data types like uint32.
---
> # These objects are plain structs like GtkTextIter containing only findamental data types like uint32.
2491,2494d2486
<   # https://github.com/StefanSalewski/gintro/issues/176
<   if gBaseInfoGetName(info) == "Rectangle" and moduleNamespace == "cairo":
<     cairoProvidesRectangle = true
< 
3822,3824c3814
<           # if not ((gStructInfoIsGtypeStruct(info) and n != "ObjectClass")): # v0.9.4 with Debian Buster crash
<           # if (not gStructInfoIsGtypeStruct(info) or (n == "ObjectClass")): # equivalent to above
<           if (not gStructInfoIsGtypeStruct(info) and (n != "ObjectClass")): # this may work -- the same as suggested by Mr. 1jss 
---
>           if not ((gStructInfoIsGtypeStruct(info) and n != "ObjectClass")):
3827c3817
<               if gStructInfoFindMethod(info, n) != nil: # crash for legacy Debian Buster
---
>               if gStructInfoFindMethod(info, n) != nil:
4108,4119c4098
<     if not cairoProvidesRectangle:
<       output.writeLine("""
< type
<   Rectangle* {.pure, byRef.} = object
<     x*: cdouble
<     y*: cdouble
<     width*: cdouble
<     height*: cdouble
< 
< proc cairo_gobject_rectangle_get_type*(): GType {.importc, libprag.}
< """)
<     output.write("\ninclude cairoimpl\n")
---
>     output.write("include cairoimpl\n")
4342c4321
< # 4342 lines
---
> # 4321 lines

```

**Comment by 1jss** [comment]
> > our files are identical except one type declaration
> 
> That is interesting, I would have expected more missing data types. If you would use an even older GTK version more types would be missing. But when that rectangle is the only one, we may try to apply that data type manually and hope that no one is using an even older GTK.
> 
> What is available by gobject-introspection is determined by the so called GIR XML files or their binary typelib versions. You can inspect the GIR files with an text editor:

I can now confirm that the type declaration for "Rectangle" is missing in `/usr/share/gir-1.0/cairo-1.0.gir `. Grep for "Rectangle" gives me just the `RectangleInt` type:

    <record name="RectangleInt" c:type="cairo_rectangle_int_t"
	    glib:type-name="CairoRectangleInt"
	    glib:get-type="cairo_gobject_rectangle_int_get_type">
      <field name="x" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="y" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="width" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
      <field name="height" writable="1">
	<type name="gint" c:type="gint"/>
      </field>
    </record>

It thus makes sense to add Rectangle manually.

**Comment by 1jss** [comment]
Great job! I will now close this issue as all the issues are resolved.
